cryptol-saw-core/src/Verifier/SAW/Cryptol.hs

```diff
@@ -1953,38 +2031,518 @@ exportRecordValue fields v =
+      --
+      --    N.B.: by using ExtCns we can write SAWCore in (`t`) that
+      --    need not keep track of de Bruijn's for the type parameters.
+      addTypeAbstractions :: Term -> IO Term
+      addTypeAbstractions t = scAbstractExts sc tyParamsECs t
+
+  -------------------------------------------------------------
+  -- Common naming conventions:
+  let sumTy_ident = identOfEnumType ntName'
+      case_ident  = identOfEnumCase ntName'
+      tl_ident    = newIdent ntName' "__LS"
+                    -- name for the 'type list', type is ListSort (thus LS)
+
+  -------------------------------------------------------------
+  -- Definitions to access needed SAWCore Prelude types & definitions:
+  sort0          <- scSort sc (mkSort 0)
+  scListSort     <- scDataTypeApp sc "Prelude.ListSort" []
+  scLS_Nil       <- scCtorApp sc "Prelude.LS_Nil"  []
+
+  let scLS_Cons s ls   = scCtorApp     sc "Prelude.LS_Cons" [s,ls]
+
+      scEithersV ls    = scGlobalApply sc "Prelude.EithersV" [ls]
+      sc_eithersV b ls = scGlobalApply sc "Prelude.eithersV" [b,ls]
+
+     -- to create values of the Either type:
+      scLeft  a b x    = scCtorApp     sc "Prelude.Left"  [a,b,x]
+      scRight a b x    = scCtorApp     sc "Prelude.Right" [a,b,x]
+
+      scMakeListSort :: [Term] -> IO Term
+      scMakeListSort = Fold.foldrM scLS_Cons scLS_Nil
+
+      -- | scMakeFunsTo b tvs - create FunsTo list of type `FunsTo b`, given
+      --    the list of (type, val :: type->b) pairs
+      scMakeFunsTo :: Term -> [(Term,Term)] -> IO Term
+      scMakeFunsTo b tvs =
+        do
+        scFunsTo_Nil <- scCtorApp sc "Prelude.FunsTo_Nil"  [b]
+        let scFunsTo_Cons (t,v) r =
+              scCtorApp sc "Prelude.FunsTo_Cons" [b,t,v,r]
+        Fold.foldrM scFunsTo_Cons scFunsTo_Nil tvs
+
+  -------------------------------------------------------------
+  -- Create TypeList(tl) for the Enum, add to SAWCore environment:
+  --  - elements of list are the elements of the Sum.
+  --  - the types maintain the same exact type vars (see tyParamsInfo)
+  -- Using the running example, we want to add the following to the
+  -- SAWCore environment:
+  --
+  --   > ETT__LS : sort 0 -> ListSort;
+  --   > ETT__LS as = LS_Cons     (scTupleType [])
+  --   >               (LS_Cons  (scTupleType [Nat])
+  --   >                (LS_Cons (scTupleType [Bool,as])
+  --   >                  LS_Nil));
+
+  --  cheating a little in the above syntax.
+  --   - scTupleType is not sawcore, it represents what's created with `scTupleType' function
+  --   - using list syntax for th ListSort lists that are the arguments to the above.
+
+  -- argTypes_eachCtor is the sum of products matrix for our enumtype:
+  (argTypes_eachCtor :: [[Term]]) <-
+    -- add tyParamsInfo to env as it is needed to allow `importType`
+    -- to work:
+    do
+    env' <- Fold.foldrM (\tp env'-> bindTParam sc tp env') env tyParamsInfo
+    flip mapM ctors $ \c->
```

Comment:
`flip mapM` is just `forM` from `Control.Monad`. I'd favor using `forM` instead. (Similar comments apply to other uses of `flip mapM` in this PR.)